base/array.jl

```diff
@@ -2327,6 +2327,29 @@ julia> filter(isodd, a)
 """
 filter(f, As::AbstractArray) = As[map(f, As)::AbstractArray{Bool}]
 
+function filter(f, a::Vector{T}) where T
+    j = 1
+    b = Vector{T}(undef, length(a))
+    for ai in a
+        @inbounds b[j] = ai
+        j = ifelse(f(ai), j+1, j)
```

Comment:
I would guess LLVM is smart enough to eliminate the branch if you write it like this?
```suggestion
        j = f(ai) ? j+1: j
```

Reply:
LLVM should. You can also use the rules of arithmetic to write it as:
```suggestion
        j += f(ai)::Bool
```

Reply:
Regarding the `j += f(ai)::Bool` variant: That works, but appears to be no faster and sometimes sligthly slower. My naive guess would be that `f(ai)` is typically encoded in a flag that can be directly used for a conditional move, and converting the flag to a 0/1 integer register plus addition takes longer (even though llvm should know that both are semantically the same and should select the fastest version regardless of what I wrote). Also, there used to be cases where explicit typeasserts confused the compiler, even if it can correctly infer that `f(ai)::Bool`. But I may well be hallucinating issues where there are none; should I still change it?

Regarding the `j = f(ai) ? j+1 : j` spelling: That becomes the same compiled code as `felse(f(ai), j+1, j)`. I just thought that the `ifelse` was more informative to the reader, since we explicitly want a conditional move instead of a branch. With this explanation, do you still prefer the `?:` spelling? Then I'll change it.